ntoskrnl/se/sid.c

```diff
@@ -587,20 +641,50 @@ SeCaptureSidAndAttributesArray(
             /* Loop the array elements */
             for (i = 0; i < AttributeCount; i++)
             {
-                /* Get the SID and it's length */
-                Sid = SrcSidAndAttributes[i].Sid;
-                SidLength = RtlLengthRequiredSid(Sid->SubAuthorityCount);
+                /*
+                 * Get the SID length from the subauthority
+                 * count we've captured before.
+                 */
+                SidLength = RtlLengthRequiredSid(ValidateArray[i].SubAuthorityCount);
 
                 /* Copy attributes */
                 SidAndAttributes[i].Attributes = SrcSidAndAttributes[i].Attributes;
 
                 /* Copy the SID to the current destination address */
                 SidAndAttributes[i].Sid = (PSID)CurrentDest;
-                RtlCopyMemory(CurrentDest, SrcSidAndAttributes[i].Sid, SidLength);
+                RtlCopyMemory(CurrentDest, ValidateArray[i].ProbeSid, SidLength);
 
-                /* Sanity checks */
-                ASSERT(RtlLengthSid(SidAndAttributes[i].Sid) == SidLength);
-                ASSERT(RtlValidSid(SidAndAttributes[i].Sid));
+                /* Obtain the SID for validation */
+                Sid = SrcSidAndAttributes[i].Sid;
```

Comment:
You want `SidAndAttributes` here, not Src -- we have to validate the data that we actually copied to kernel memory; since the UM memory may still change under us after the `RtlCopyMemory`